cuda/matrix/dense_kernels.cu

```diff
@@ -346,20 +349,199 @@ void compute_norm2(std::shared_ptr<const CudaExecutor> exec,
+    size_type num_rows, size_type work_size,
+    size_type *__restrict__ nnz_per_row, size_type *__restrict__ add_values)
+{
+    const auto tidx = threadIdx.x + blockDim.x * blockIdx.x;
+    if (tidx < num_rows) {
+        auto start_row = tidx * work_size;
+        auto nnz_up_to_here = nnz_per_row[start_row];
+        for (auto i = start_row + 1; i < min(start_row + work_size, num_rows);
+             i++) {
+            nnz_up_to_here += nnz_per_row[i];
+            nnz_per_row[i] = nnz_up_to_here;
+        }
+        add_values[tidx] = nnz_up_to_here;
+    }
+}
+
+__global__ __launch_bounds__(default_block_size) void finalize_nnz_sum(
+    size_type num_rows, size_type *__restrict__ nnz_per_row,
+    size_type *__restrict__ add_values)
+{
+    const auto tidx = threadIdx.x + blockIdx.x * blockDim.x;
+
+    if (tidx + blockDim.x < num_rows) {
+        size_type add_value = 0;
+        for (auto i = 0; i <= blockIdx.x; i++) {
+            add_value += add_values[i];
+        }
+        nnz_per_row[tidx + blockDim.x] += add_value;
+    }
+}
+
+template <typename ValueType, typename IndexType>
+__global__ __launch_bounds__(default_block_size) void fill_in_coo(
+    size_type num_rows, size_type num_cols, size_type stride,
+    const size_type *__restrict__ row_ptrs,
+    const ValueType *__restrict__ source, IndexType *__restrict__ row_idxs,
+    IndexType *__restrict__ col_idxs, ValueType *__restrict__ values)
+{
+    const auto tidx = threadIdx.x + blockDim.x * blockIdx.x;
+    if (tidx < num_rows) {
+        size_type write_to = (tidx == 0) ? 0 : row_ptrs[tidx - 1];
+
+        for (auto i = 0; i < num_cols; i++) {
+            if (source[stride * tidx + i] != zero<ValueType>()) {
+                values[write_to] = source[stride * tidx + i];
+                col_idxs[write_to] = i;
+                row_idxs[write_to] = tidx;
+                write_to++;
+            }
+        }
+    }
+}
+
+}  // namespace kernel
 
 template <typename ValueType, typename IndexType>
 void convert_to_coo(std::shared_ptr<const CudaExecutor> exec,
                     matrix::Coo<ValueType, IndexType> *result,
-                    const matrix::Dense<ValueType> *source) NOT_IMPLEMENTED;
+                    const matrix::Dense<ValueType> *source)
+{
+    auto num_rows = result->get_size()[0];
+    auto num_cols = result->get_size()[1];
+
+    auto row_idxs = result->get_row_idxs();
+    auto col_idxs = result->get_col_idxs();
+    auto values = result->get_values();
+
+    auto stride = source->get_stride();
+
+    auto nnz_per_row = Array<size_type>(exec);
+    nnz_per_row.resize_and_reset(num_rows);
+    calculate_nonzeros_per_row(exec, source, &nnz_per_row);
+
+    const size_type grid_dim = ceildiv(num_rows, default_block_size);
+    auto add_values = Array<size_type>(exec);
+    add_values.resize_and_reset(grid_dim);
+
+    kernel::start_nnz_sum<<<ceildiv(grid_dim, cuda_config::warp_size),
```

Comment:
Each thread gets assigned a chunk of default_block_size elements, and there are grid_dim of those chunks. The grid_dim is actually the grid dimension for the next kernel.

Reply:
You are correct, seems I got confused there for a bit (because the workload per thread is `default_block_size`).

However, I would prefer if you would use the prefix computation here (less redundant code and more performance).